vault/login_mfa.go

```diff
@@ -2013,11 +2015,29 @@ func (c *Core) validateTOTP(ctx context.Context, creds []string, entityMethodSec
 
 	usedName := fmt.Sprintf("%s_%s", configID, creds[0])
 
-	_, ok := c.loginMFABackend.usedCodes.Get(usedName)
+	_, ok := usedCodes.Get(usedName)
 	if ok {
 		return fmt.Errorf("code already used; new code is available in %v seconds", totpSecret.Period)
 	}
 
+	numAttempts, _ := usedCodes.Get(configID)
+	if numAttempts == nil {
+		usedCodes.Set(configID, int64(1), defaultMFAAuthResponseTTL)
+	} else {
+		err := usedCodes.Increment(configID, 1)
+		if err != nil {
+			return fmt.Errorf("failed to increment the TOTP code counter")
+		}
+	}
+	numAttempts, _ = usedCodes.Get(configID)
+	num, ok := numAttempts.(int64)
+	if !ok {
+		return fmt.Errorf("invalid counter type returned in TOTP usedCode cache")
+	}
+	if num > maximumValidationAttempts {
+		return fmt.Errorf("maximum TOTP validation attempts %d exceeded %d", num, maximumValidationAttempts)
+	}
```

Comment:
It feels like you could move this block, starting on line 2033, inside the `else` statement on line 2026. If `numAttempts` is nil, then clearly we're not in danger of exceeding the maximum validation attempts. That would avoid the need for the redundant `usedCodes.Get(configID)` call on line 2032.

Reply:
Sounds great!